theseus/optimizer/nonlinear/dcem.py

```diff
@@ -0,0 +1,245 @@
+from typing import Callable, NoReturn, Optional, Union, List
+
+import numpy as np
+import torch
+from torch.distributions import Normal
+
+from theseus.third_party.lml import LML
+from theseus.core.objective import Objective
+from theseus.optimizer import OptimizerInfo
+from theseus.optimizer.variable_ordering import VariableOrdering
+
+from .nonlinear_optimizer import (
+    NonlinearOptimizer,
+    BackwardMode,
+    NonlinearOptimizerInfo,
+    NonlinearOptimizerStatus,
+)
+
+EndIterCallbackType = Callable[
+    ["DCEM", NonlinearOptimizerInfo, torch.Tensor, int], NoReturn
+]
+
+
+class DCEM(NonlinearOptimizer):
+    """
+    DCEM optimizer for nonlinear optimization using sampling based techniques.
+    The optimizer can be really sensitive to hypermeter tuning. Here are few tuning
+    hints:
+    1. If have to lower the max_iterations, then increase the n_sample.
+    2. The higher the n_sample, the slowly with variance of samples will decrease.
+    3. The higher the n_sample, more the chances of optimum being in the elite set.
+    4. The higher the n_elite, the slower is convergence, but more accurate it might
+       be, but would need more iterations. n_elite= 5 is good enough for most cases.
+    """
+
+    def __init__(
+        self,
+        objective: Objective,
+        vectorize: bool = False,
+        max_iterations: int = 50,  # 50 for test_theseus
+        n_sample: int = 100,  # 100 for test_theseus
+        n_elite: int = 5,  # 5
+        temp: float = 1.0,
+        init_sigma: Union[float, torch.Tensor, List[float]] = 1.0,
+        lb=None,
+        ub=None,
+        lml_verbose: bool = False,
+        lml_eps: float = 1e-3,
+        normalize: bool = True,
+        abs_err_tolerance: float = 1e-6,
+        rel_err_tolerance: float = 1e-4,
+        **kwargs,
+    ) -> None:
+        super().__init__(
+            objective,
+            vectorize=vectorize,
+            abs_err_tolerance=abs_err_tolerance,
+            rel_err_tolerance=rel_err_tolerance,
+            max_iterations=max_iterations,
+            **kwargs,
+        )
+
+        self.objective = objective
+        self.ordering = VariableOrdering(objective)
+        self.n_samples = n_sample
+        self.n_elite = n_elite
+        self.lb = lb
+        self.ub = ub
+        self.temp = temp
+        self.normalize = normalize
+        self.tot_dof = sum([x.dof() for x in self.ordering])
+        self.lml_eps = lml_eps
+        self.lml_verbose = lml_verbose
+        self.init_sigma = init_sigma
+
```

Comment:
Yeah, I guess. But info seems to have specific attributes about convergence, errors, history, etc. And since other optimizer also use info class and they don't anything returning like sigma, I am not sure if info is the right place? On the other hand, info makes sense, but maybe adding new attribute such as `auxiliary_info` which is empty by default, but optimizer can add any info they want to return through this? 
May be can keep this for different PR.